designs/0004-sparse-matrices.md

```diff
@@ -0,0 +1,387 @@
+
+## Full Example Model
+
+Below is a full example model that uses sparse vectors and matrices for data and parameters for a gaussian process.
+
+```stan
+data {
+  int<lower=1> N; // Vec size
+  int<lower=1, upper=N> K; // # of nonzero elements
+  int<lower=1> non_zero_index[K]; // Indices for nonzero elements
+  sparse_vector[N, non_zero_index] x;
+  vector[N] y;
+}
+transformed data {
+  sparse_vector[N] mu = rep_sparse_vector(0, N, non_zero_index); // [1]
+}
+parameters {
+  real<lower=0> rho;
+  real<lower=0> alpha;
+  real<lower=0> sigma;
+}
+model {
+  sparse_cholesky_matrix[N] L_K; // [2]
+  sparse_matrix[N, N] K = cov_exp_quad(x, alpha, rho);
+  real sq_sigma = square(sigma);
+
+  // Assign to pre-existing diagonal
+  for (n in 1:N)
+    K[n, n] = K[n, n] + sq_sigma;
+
+  L_K = cholesky_decompose(K);
+
+  rho ~ inv_gamma(5, 5);
+  alpha ~ std_normal();
+  sigma ~ std_normal();
+
+  y ~ multi_normal_cholesky(mu, L_K);
+}
+```
+
+- [1] Because Stan is told what values are true zeros vs. sparse zeros we can set this up fine.
+
+- [2] The `sparse_cholesky_matrix` is optional and holds the permutation of the sparse matrix after decomposition using the `EIGEN_SPARSEMATRIX_BASE_PLUGIN`. Permuting the rows of the sparse matrix before decomposition can often increase performance. Instead of having to recompute the permutation on each call to `cholesky_decompose`, after the first iteration we can keep the [`PermutationMatrix`](https://eigen.tuxfamily.org/dox/classEigen_1_1SimplicialCholeskyBase.html#aff1480e595a21726beaec9a586a94d5a) inside of the `sparse_cholesky_matrix` and reuse it on subsequent operations.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## I/O
+
+Data input formats should not need to change, less R and Python want a particular method that does not exist at this time.
+
+At the C++ level input data can be constructed to the sparse matrix through soemthing like what Ben did [here](https://github.com/stan-dev/rstanarm/blob/master/inst/include/csr_matrix_times_vector2.hpp#L18) for a revised `csr_matrix_times_vector`.
+
+```cpp
+int outerIndexPtr[cols+1];
+int innerIndices[nnz];
+double values[nnz];
+// read-write (parameters)
+Map<SparseMatrix<double>> sm1(rows, cols, nnz, outerIndexPtr, innerIndices, values);
+// read only (data)
+Map<const SparseMatrix<double>> sm2(rows, cols, nnz, outerIndexPtr, innerIndices, values);   
+```
+
+## Stan Math
+
+### Templating
+
+There are two approaches to have Stan support Eigen's Sparse Matrix format.
+
+#### The Hard way
+
+Sparse matrices can be supported in Stan-math by either moving to `EigenBase` as the default in the metaprogramming or by having separate methods for Sparse Matrices.
```

Comment:
Moving to EigenBase (or whichever) would be really exciting and probably deserve its own like, forum post as a proposal (at least to talk through the steps to achieve it and which base class to use before coding).

I kinda think that's probably the route we wanna go, which makes me wonder about the timeliness of this proposal, haha. Should it be staged after we switch?

Reply:
> Moving to EigenBase (or whichever) would be really exciting and probably deserve its own like, forum post as a proposal (at least to talk through the steps to achieve it and which base class to use before coding).

> I kinda think that's probably the route we wanna go, which makes me wonder about the timeliness of this proposal, haha. Should it be staged after we switch?

Maybe, I think I can start working on the Stan math parts that won't use EigenBase like `inverse` and `cholesky_decompose` as well as the rev implimentation

Reply:
Also it looks like the example above causes to mov and copy constructors to go off so maybe it's nada

https://godbolt.org/z/1z6KMp